feat: Unified REST API

lostbean opened the pull request:
## Description:
This PR creates unified and first class REST API to manage the Kurtosis cluster (enclaves, services, ...). The REST APIs are defined using OpenAPI 3 specs, oapi-codegen to generate server and golang bindings and echo to server it. It also server Swagger UI to allow exploring the current API documentation.

!!! WARING !!! This PR also include bumping the Go compiler to `1.20` due requirement in some of the new dependencies for OpenAPI specs.

## Is this change user facing?
No
<!-- If yes, please add the "user facing" label to the PR -->
<!-- If yes, don't forget to include docs changes where relevant -->

## References (if applicable):
<!-- Add relevant Github Issues, Discord threads, or other helpful information. -->


BEGIN_COMMIT_OVERRIDE
feat: Unified REST API
END_COMMIT_OVERRIDE

lostbean committed:
add initial http api

lostbean committed:
make it restful

lostbean committed:
add generated golang bindings code for new REST api

lostbean committed:
use Echo framework

lostbean committed:
add test server

lostbean committed:
Add strict interface

lostbean committed:
Decouple inner types

lostbean committed:
add rest http server to engine

lostbean committed:
tidy cli mod

lostbean committed:
allow multiple shells on ktdev completion

lostbean committed:
Do actual registration of the handlers

lostbean committed:
fix pointers

lostbean committed:
return 404 on enclave uuid

lostbean committed:
log error from log channel

lostbean committed:
rename file

lostbean committed:
bind new REST API port to engine container

lostbean committed:
clean up test service

lostbean committed:
add enclave id to APIC's REST api path

lostbean committed:
start enclave management REST API on engine

lostbean committed:
add artifact upload

lostbean committed:
add upload of remote files

lostbean committed:
Use streaming utils to manage local file upload

lostbean committed:
Pass remote upload to APIC

lostbean committed:
add API to store service file

lostbean committed:
Normalize artifact references

lostbean committed:
add artifact inspection

lostbean committed:
add artifact download

lostbean committed:
Use streaming to download artifacts

lostbean committed:
list services

lostbean committed:
add Connect command

lostbean committed:
Get service info

lostbean committed:
Add command execution

lostbean committed:
Add availability checker

lostbean committed:
Fetch info about last Starlark run

lostbean committed:
Add package upload

lostbean committed:
Run both Starlark scripts and packages

lostbean committed:
Add Swagger UI endpoints to visualize deployed API

lostbean committed:
Remove non-string args from log msg

lostbean committed:
rename files

lostbean committed:
rename schemas on engine API

lostbean committed:
more renaming on schemas on engine API

lostbean committed:
Update optionality on engine API schemas

lostbean committed:
use utils functions

lostbean committed:
rename schemas on enclave API

lostbean committed:
Update optionality on core API schemas

lostbean committed:
switch between internal and host connection for APIC

lostbean committed:
Add websocket for logs

lostbean committed:
remove duplicated parameter

lostbean committed:
handle dead enclave/missing APIC

lostbean committed:
fix package run

lostbean committed:
rename with shorter namespace

lostbean committed:
more restful

lostbean committed:
move availability check to GET

lostbean committed:
add grpcui to test gRPC calls

lostbean committed:
clean file for old protobuf to swagger conversion

lostbean committed:
handle enclave/engine not found

lostbean committed:
clean-up unused code

lostbean committed:
bump go version on CI

omar711 commented on api/openapi/engine/engine_service.yaml:
```diff
@@ -0,0 +1,337 @@
+        - creation_time
+        - mode
+
+    EnclaveAPIContainerInfo:
```
In parallel I've added a couple of fields to `EnclaveAPIContainerInfo` and `EnclaveAPIContainerHostMachineInfo` to expose the tunnel server's ports. 
Reference is here: https://github.com/kurtosis-tech/kurtosis/pull/1849/files#diff-a24e12c03e41a9365ffc8601baddeee397de4b5547611039da5e15ac24d82a22R106 

Not sure on best timing, but we could add these fields right away, make them required and populate them with zeros to begin with?  Specifically this would be fields:

* `EnclaveAPIContainerInfo.tunnel_port_inside_enclave (integer, required)`
* `EnclaveAPIContainerHostMachineInfo.tunnel_port_on_host_machine (integer, required)`

which should avoid transitions at a later date

lostbean replied:
Thanks for the heads up on these new fields but since they are additives we can also add them later without any break changes.

leoporoli commented on .circleci/config.yml:
```diff
@@ -16,7 +16,7 @@ param_cli_cluster_backend: &param_cli_cluster_backend
     type: enum
     enum: ["docker", "kubernetes"]
 
-# Install go version 1.19 to run our integration tests on the minimal version we support
```
Make sure that you merge and release these changes on a breaking change day.

leoporoli commented on api/golang/core/kurtosis_core_rest_api_bindings/api_container_server.gen.go:
TS bindings are missing, or do you plan to add them later?
What will happen with the Rust SDK? Will it be supported?

lostbean replied:
I'm planning to add the TS/Go bindings on a second PR.

Regarding the Rust SDK, didn't plan to include that. @leeederek / @h4ck3rk3y should I include a migration of RUST SDK now? Do we have active users relying on it?

h4ck3rk3y replied:
i went searching on GitHub

https://github.com/search?q=-org%3Akurtosis-tech+kurtosis-sdk&type=code

i can't see it in a `cargo.toml` thats not a Fork of Kurtosis; LayerZero has a fork that they committed to in October; but not sure if they use rust/the fork

lostbean replied:
Cool, I'll skip it for now but we should decide to drop it or migrate it before the code clean after we are planning for later.

leoporoli commented on api/openapi/core/server.cfg.yaml:
```diff
@@ -0,0 +1,6 @@
+  embedded-spec: true
+  echo-server: true
+  strict-server: true
+output: api/golang/core/kurtosis_core_rest_api_bindings/api_container_server.gen.go
```
nit: end line missing

leoporoli commented on api/openapi/core/types.cfg.yaml:
```diff
@@ -0,0 +1,4 @@
+package: kurtosis_core_rest_api_bindings
+generate:
+  models: true
+output: api/golang/core/kurtosis_core_rest_api_bindings/api_container_types.gen.go
```
nit: end line missing

lostbean replied:
done

leoporoli commented on api/openapi/engine/engine_service.yaml:
```diff
@@ -0,0 +1,337 @@
+openapi: 3.0.0
+
+info:
+  title: Engine API
```
WDYTA specifying that it is a REST API?

leoporoli commented on api/openapi/engine/engine_service.yaml:
```diff
@@ -0,0 +1,337 @@
+
+info:
+  title: Engine API
+  version: 0.1.0
```
for acquiring knowledge. When do we have to bump this version? when we make any small change in the API right? what happens if, for instance, we add a 500 response line in an endpoint, do we have to bump the version in that case?

lostbean replied:
We should use https://semver.org/ here. So adding a 500 response would be a PATCH bump

lostbean committed:
Add pointer null checks

lostbean committed:
Implement get info handler

lostbean committed:
Rename spec

lostbean committed:
Split and merge API types

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+      parameters:
+        - $ref: './api_types.yaml#/components/parameters/enclave_identifier'
+      requestBody:
+        $ref: "./api_types.yaml#/components/requestBodies/fileUploadBody"
```
I don't see the fileUploadBody inside the "api_types.yaml" Do I missing something?

leoporoli replied:
is this a stream right? so no type definition I guess

lostbean replied:
Ended up merge all specs into a single file. But the definition is now [here](https://github.com/kurtosis-tech/kurtosis/blob/f3cb0c0d363bed2402603420b7e120d146980482/api/openapi/specs/kurtosis_api.yaml#L597)

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+
+  /enclaves/{enclave_identifier}/starlark/packages/{package_id}:
+    post:
+      description: Executes a Starlark script on the user's behalf
```
```suggestion
      description: Executes a Starlark package on the user's behalf
```

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+                items:
+                  $ref: "./api_types.yaml#/components/schemas/StarlarkRunResponseLine"
+
+  /enclaves/{enclave_identifier}/starlark/scripts:
```
suggestion, it could be also "/enclaves/{enclave_identifier}/starlark/module" because I single file script is a module and several modules compose a package

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+
+  /enclaves/{enclave_identifier}/starlark/scripts:
+    post:
+      description: Executes a Starlark script on the user's behalf
```
it can also be "module"

leoporoli replied:
mmm but now thinking about it, the methods in the backend are using "script" so yeap! maybe we should continue with that pattern

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+                items:
+                  $ref: "./api_types.yaml#/components/schemas/StarlarkRunResponseLine"
+
+  /enclaves/{enclave_identifier}/services/{service_identifier}:
```
missing description

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+                items:
+                  $ref: "./api_types.yaml#/components/schemas/ServiceIdentifiers"
+
+  /enclaves/{enclave_identifier}/services:
```
missing description

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+
+  /enclaves/{enclave_identifier}/services/{service_identifier}/command:
+    post:
+      description: Executes the given command inside a running container
```
```suggestion
      description: Executes the given command inside a running service's container
```

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+        - $ref: './api_types.yaml#/components/parameters/service_identifier'
+        - $ref: './api_types.yaml#/components/parameters/port_number'
+        - $ref: './api_types.yaml#/components/parameters/http_method'
+        - $ref: './api_types.yaml#/components/parameters/path'
```
some of these parameters are query params right? should they be specified here?

lostbean replied:
Yes, they are specified in the `in:` field ([example](https://github.com/kurtosis-tech/kurtosis/blob/f3cb0c0d363bed2402603420b7e120d146980482/api/openapi/specs/kurtosis_api.yaml#L665))

leoporoli commented on api/openapi/specs/core_service.yaml:
```diff
@@ -0,0 +1,331 @@
+        - $ref: './api_types.yaml#/components/parameters/retries'
+        - $ref: './api_types.yaml#/components/parameters/retries_delay_milliseconds'
+        - $ref: './api_types.yaml#/components/parameters/expected_response'
+        - $ref: './api_types.yaml#/components/parameters/request_body'
```
my only concern here is that GET request has char limit I think, and the request body could be long (in general should be small, but for corner cases), but probably is ok for now and if we see users having this problem we can move it to a POST request maybe

lostbean replied:
Fair point! I think the limit is 2048 but most of these fields are optional and we don't expect them to any large since those are for status checking endpoints. From the RESTful perspective, and GET makes more sense since they have no side effect in the backend. Let's leave as a GET for now, and we can reconsider it if we actually start runinig into issues.

leoporoli commented on api/openapi/specs/engine_service.yaml:
```diff
@@ -0,0 +1,149 @@
+          content:
+            application/json:
+              schema:
+                type: object
```
shouldn't it return an array?

lostbean replied:
Not an array but is returning dictionary map[string]enclaveInfo

leoporoli commented on api/openapi/specs/engine_service.yaml:
```diff
@@ -0,0 +1,149 @@
+            schema:
+              $ref: './api_types.yaml#/components/schemas/CreateEnclave'
+      responses:
+
```
nit: empty line

leoporoli commented on api/openapi/specs/engine_service.yaml:
```diff
@@ -0,0 +1,149 @@
+    delete:
+      summary: Delete Enclaves
+      parameters:
+        - $ref: './api_types.yaml#/components/parameters/remove_all'
```
probably the description could describe what happens when the value is false, I think only stopped enclave will be removed, right?

lostbean replied:
yes!

leoporoli commented on api/openapi/specs/engine_service.yaml:
```diff
@@ -0,0 +1,149 @@
+          content:
+            application/json:
+              schema:
+                $ref: './api_types.yaml#/components/schemas/EnclaveContainersStatus'
```
could it be just EnclaveStatus?

lostbean replied:
+1

leoporoli commented on api/openapi/specs/websocket_service.yaml:
```diff
@@ -0,0 +1,42 @@
+      summary: Get Service Logs
+      parameters:
+        - $ref: "./api_types.yaml#/components/parameters/enclave_identifier"
+        - $ref: "./api_types.yaml#/components/parameters/service_uuid_set"
```
some of these are query params right?

lostbean replied:
yes, the ones that are not in the path

leoporoli commented on api/openapi/specs/websocket_service.yaml:
```diff
@@ -0,0 +1,42 @@
+paths:
+  /enclaves/{enclave_identifier}/logs:
+    get:
+      summary: Get Service Logs
```
```suggestion
      summary: Get Enclave's Services Logs
```

leoporoli commented on api/openapi/specs/websocket_service.yaml:
```diff
@@ -0,0 +1,42 @@
+      parameters:
+        - $ref: "./api_types.yaml#/components/parameters/enclave_identifier"
+        - $ref: "./api_types.yaml#/components/parameters/service_identifier"
+        - $ref: "./api_types.yaml#/components/parameters/follow_logs"
```
some of these are query params right?

leoporoli commented on container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/create_engine.go:
```diff
@@ -17,13 +19,13 @@ import (
 	//TODO: pass this parameter
 	enclaveManagerUIPort                        = 9711
 	enclaveManagerAPIPort                       = 8081
+	restAPIPort                                 = 9779 //TODO: pass this parameter
```
will you make this change in the next PR?

lostbean replied:
Ok, I decided indeed to move it to the clean-up phase, since of this ports/args might go away.

leoporoli commented on container-engine-lib/lib/backend_impls/docker/docker_kurtosis_backend/engine_functions/create_engine.go:
```diff
@@ -134,6 +136,16 @@ func CreateEngine(
+	if err != nil {
+		return nil, stacktrace.Propagate(
+			err,
+			"An error occurred creating the Enclave Manager UI's http port spec object using number '%v' and protocol '%v'",
```
is this the enclave manager or the engine's rest api?

lostbean replied:
yes!

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -60,6 +69,8 @@ const (
 
 	shouldFlushMetricsClientQueueOnEachEvent = false
+
+	restAPIPortAddr uint16 = 9779
```
it should be received on the serverArgs

lostbean replied:
Same here, I decided indeed to move it to the clean-up phase, since of this ports/args might go away.

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	logrus.Info("Running REST API server...")
+
+	// This is how you set up a basic Echo router
+	e := echo.New()
```
I suggest using semantic var names, we try to not use one char var name.

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	// OpenAPI schema.
+	swagger_engine, err := engineApi.GetSwagger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
```
can we replace this with a Logrus function?

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+
+	// ============================== Engine Management API ======================================
+	// OpenAPI schema.
+	swagger_engine, err := engineApi.GetSwagger()
```
I suggest to change the var name to swaggerEngine

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	swagger_engine, err := engineApi.GetSwagger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
+		os.Exit(1)
```
what will happen in this case? will the other engine servers gracefully shut down?

adschwartz replied:
Agreed, they won't shut down. I've mentioned this before, but we should really not be running our webservers in the same containers as the engine itself. It's messy at best to handle the loss of a process inside the container. The best practice is to launch them as separate containers and use a healthcheck to detect the issue.
@lostbean

lostbean replied:
I haven't put a lot of thought on this part of the code but you are both correct. I've improved and mitigated it for now (skipping errors from swagger since it's now a critical part of the engine). The swagger spec could move together with the Web Server (static asset server which btw is also [running from the engine](https://github.com/kurtosis-tech/kurtosis/blob/0c231668df9357a02b209118a5aa7f1eb85c1f24/engine/server/engine/main.go#L158)). I've put a  TODO to move it once we enter the clean-up and split the web server into its own service.

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+
+	// ============================== Logging API ======================================
+
+	swagger_websocket, err := loggingApi.GetSwagger()
```
I suggest to change the var name to swaggerWebsocket

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+
+	swagger_websocket, err := loggingApi.GetSwagger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
```
same comments here for using logrus

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
+		os.Exit(1)
+	}
+	server.ServeSwaggerUI(e, "/api/specs/engine", server.NewSwaggerUIConfig(swagger_engine))
```
suggest to use a constant for the second argument

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
+		os.Exit(1)
+	}
+	server.ServeSwaggerUI(e, "/api/specs/websocket", server.NewSwaggerUIConfig(swagger_websocket))
```
suggest to use a constant for the second argument

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		LogFileManager:              logFileManager,
+		MetricsClient:               metricsClient,
+	}
+	// TODO(edgar) add logging Close()
```
is this TODO for the next PR's? just calling in case you forgot.

lostbean replied:
implementation has changed but it's done

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		MetricsClient:               metricsClient,
+	}
+	// TODO(edgar) add logging Close()
+	// defer webSocketRuntime.ShutDown()
```
it should be activated, right?

lostbean replied:
implementation has changed but it's done

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+
+	// ============================== Engine Management API ======================================
+	// OpenAPI schema.
+	swagger_enclave, err := enclaveApi.GetSwagger()
```
I suggest to change the var name to swaggerEnclave

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	// OpenAPI schema.
+	swagger_enclave, err := enclaveApi.GetSwagger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
```
same here for Logrus

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		fmt.Fprintf(os.Stderr, "Error loading swagger spec\n: %s", err)
+		os.Exit(1)
+	}
+	server.ServeSwaggerUI(e, "/api/specs/enclave", server.NewSwaggerUIConfig(swagger_enclave))
```
suggest to use a constant for the second argument

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	// e.Use(middleware.OapiRequestValidator(swagger_enclave))
+	enclaveRuntime, err := restApi.NewEnclaveRuntime(ctx, *enclave_manager, false)
+	if err != nil {
+		// TODO(edgar) fix error handling
```
missing propagate I think

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+		// TODO(edgar) fix error handling
+	}
+	// TODO(edgar) add enclave Close()
+	// defer enclaveRuntime.ShutDown()
```
it should be activated, right? And the enclave close maybe?

lostbean replied:
implementation has changed but it's done

leoporoli commented on engine/server/engine/main.go:
```diff
@@ -338,3 +359,87 @@ func formatFilenameFunctionForLogs(filename string, functionName string) string
+	enclaveApi.RegisterHandlers(e, enclaveApi.NewStrictHandler(enclaveRuntime, nil))
+
+	// ============================== Start Server ======================================
+	e.Logger.Fatal(e.Start(net.JoinHostPort("0.0.0.0", fmt.Sprint(restAPIPortAddr))))
```
suggest using a constant for "0.0.0.0"

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	lock                     sync.Mutex
+}
+
+func (runtime enclaveRuntime) refreshEnclaveConnections() error {
```
nit: we usually put the private functions or methods after the public ones

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	enclaves, err := runtime.enclaveManager.GetEnclaves(runtime.ctx)
+	if err != nil {
+		return err
```
missing propagate

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	}
+
+	// Clean up removed enclaves
+	for uuid := range runtime.remoteApiContainerClient {
```
what happens if this map contains a client for an enclave that was not returned by the enclave manager? shouldn't it be also deleted or not?

lostbean replied:
Yes, that is what is going to happen. It basically removes the enclaves that are not listed in the GetEnclaves after a refresh. Added more info in the comment.

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	// Add new enclaves - assuming enclaves properties (API container connection) are immutable
+	for uuid, info := range enclaves {
+		_, found := runtime.remoteApiContainerClient[uuid]
+		if !found && (info != nil) {
```
```suggestion
		if !found && info != nil {
```

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+			conn, err := getGrpcClientConn(*info, runtime.connectOnHostMachine)
+			if err != nil {
+				logrus.Errorf("Failed to establish gRPC connection with enclave manager service on enclave %s", uuid)
+				return err
```
missing propagate

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	err := runtime.refreshEnclaveConnections()
+	if err != nil {
+		return nil, err
```
missing propagate

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	artifacts, err := (*apiContainerClient).ListFilesArtifactNamesAndUuids(ctx, &emptypb.Empty{})
+	if err != nil {
+		return nil, err
```
missing propagation and I suggest to review it throughout the file

leoporoli commented on engine/server/engine/utils/list.go:
```diff
@@ -0,0 +1,21 @@
+	return res
+}
+
+func MapListStopOnError[T, U any](data []T, f func(T) (U, error)) ([]U, error) {
```
is not used

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+// (POST /enclaves/{enclave_identifier}/artifacts/local-file)
+func (manager *enclaveRuntime) PostEnclavesEnclaveIdentifierArtifactsLocalFile(ctx context.Context, request api.PostEnclavesEnclaveIdentifierArtifactsLocalFileRequestObject) (api.PostEnclavesEnclaveIdentifierArtifactsLocalFileResponseObject, error) {
+	enclave_identifier := request.EnclaveIdentifier
```
use golang var name convention, I suggest to review it throughout the file

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+		}
+		clientStream := grpc_file_streaming.NewClientStream[rpc_api.StreamedDataChunk, rpc_api.UploadFilesArtifactResponse](client)
+
+		response, err := clientStream.SendData(
```
missing err check

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+		response, err := clientStream.SendData(
+			filename,
+			part,
+			0, // Length unknown head of time
```
magic number

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+		)
+
+		// The response is nil when a file artifact with the same has already been uploaded
+		// TODO (edgar) Is this the expected behavior? If so, we should be explicit about it.
```
I think we should be explicit about it, probably a 400 status code

lostbean replied:
+1, I added a per-file response using a 409

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+		ContentLength: 0, // No file size is provided since we are streaming it directly
+	}
+
+	return api.GetEnclavesEnclaveIdentifierArtifactsArtifactIdentifierDownload200ApplicationoctetStreamResponse(response), nil
```
is not necessary to cast the var here because it's already this type

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	response := api.GetEnclavesEnclaveIdentifierArtifactsArtifactIdentifierDownload200ApplicationoctetStreamResponse{
+		Body:          pipeReader,
+		ContentLength: 0, // No file size is provided since we are streaming it directly
```
magic number

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	mapped_services := utils.MapMapValues(services.ServiceInfo, toHttpServiceInfo)
+	selected_service, found := mapped_services[service_identifier]
+	if !found {
+		// TODO(edgar) add 404 return
```
just calling if we forgot this on this PR

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	jsonParams := utils.DerefWith(request.Body.Params, map[string]interface{}{})
+	jsonBlob, err := json.Marshal(jsonParams)
+	if err != nil {
+		panic("Failed to serialize parameters")
```
shouldn't be better to return a request error or server error instead of panicking here? panic will stop the server I think

lostbean replied:
IMO panicking is fine if you really don't expect that to fail, which is the case here. We are panicking here because we are serializing a deserialized JSON parameter and that never should fail. Also panicking here doesn't bring the server down, it only brings crashes the go routine handling that HTTP request. That said, I don't see any big issue on adding a bit of more code to pass the error back to the handler.

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	// TODO: Why the gRPC is not sending the previous hash??? and we need to calc it manually
+	hasher := sha1.New()
+	previousBlockHash := ""
```
you can replace these with constants

lostbean replied:
this piece of code is now gone

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	}
+	clientStream := grpc_file_streaming.NewClientStream[rpc_api.StreamedDataChunk, []byte](client)
+	pipeReader := clientStream.PipeReader(
+		"__RunStarlarkScript__",
```
I suggest to use a constant here

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	)
+	response := api.PostEnclavesEnclaveIdentifierStarlarkScripts200ApplicationoctetStreamResponse{
+		Body:          pipeReader,
+		ContentLength: 0, // No file size is provided since we are streaming it directly
```
magic numebr

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+
+	client, found := manager.remoteApiContainerClient[enclave_uuid]
+	if !found {
+		return nil, nil
```
I think it should return an error here or the caller should check for nil values

lostbean replied:
Yes, and the client [does check for nil values](https://github.com/kurtosis-tech/kurtosis/blob/07c7f8e47f0dfab5a5fabea5b64b8cde1bd21ec7/engine/server/engine/server/enclave_rest_api_handler.go#L719)

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	case api_type.NOCONNECT:
+		return rpc_api.Connect_NO_CONNECT
+	default:
+		panic(fmt.Sprintf("Missing conversion of Connect Enum value: %s", conn))
```
I suggest returning an error instead of panicking here

lostbean replied:
this is one of the panicking could make sense but we don't expect a failure under normal operations. This is an on-to-one mapping between enums and the only failure mode is if one adds a new value to the enum. I'm fine adding code to handle it throughout the codebase (as we need to keep passing it through the function stack calls) but it would be great if Go would have a native concept of enum exhaustion in switch statements (maybe we can work to add smtg like https://pkg.go.dev/github.com/nishanths/exhaustive to enforce that check).

lostbean replied:
I've added the [exhaustive](https://pkg.go.dev/github.com/nishanths/exhaustive) to our golangci-lint and the default case should never be reached.

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	case rpc_api.ServiceStatus_UNKNOWN:
+		return api_type.ServiceStatusUNKNOWN
+	default:
+		panic(fmt.Sprintf("Missing conversion of Service Status Enum value: %s", status))
```
same here

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	case rpc_api.Container_UNKNOWN:
+		return api_type.ContainerStatusUNKNOWN
+	default:
+		panic(fmt.Sprintf("Missing conversion of Container Status Enum value: %s", status))
```
same here

leoporoli commented on engine/server/engine/server/enclave_rest_api_handler.go:
```diff
@@ -0,0 +1,857 @@
+	case rpc_api.Port_SCTP:
+		return api_type.SCTP
+	default:
+		panic(fmt.Sprintf("Missing conversion of Transport Protocol Enum value: %s", protocol))
```
same here

leoporoli commented on engine/server/engine/server/engine_connect_server_service.go:
```diff
@@ -75,6 +79,94 @@ func NewEngineConnectServerService(
+	case types.EnclaveContainersStatus_RUNNING:
+		return kurtosis_engine_rpc_api_bindings.EnclaveContainersStatus_EnclaveContainersStatus_RUNNING
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", status))
```
propagate an error instead of panicking

leoporoli commented on engine/server/engine/server/engine_connect_server_service.go:
```diff
@@ -75,6 +79,94 @@ func NewEngineConnectServerService(
+	case types.ContainerStatus_RUNNING:
+		return kurtosis_engine_rpc_api_bindings.EnclaveAPIContainerStatus_EnclaveAPIContainerStatus_RUNNING
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", status))
```
same here

leoporoli commented on engine/server/engine/server/engine_connect_server_service.go:
```diff
@@ -75,6 +79,94 @@ func NewEngineConnectServerService(
+	case types.EnclaveMode_TEST:
+		return kurtosis_engine_rpc_api_bindings.EnclaveMode_TEST
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", mode))
```
same here

leoporoli commented on engine/server/engine/server/engine_rest_api_handler.go:
```diff
@@ -0,0 +1,279 @@
+	info, found := infoForEnclaves[request.EnclaveIdentifier]
+	if !found {
+		notFoundErr := stacktrace.NewError("Enclave '%s' not found.", request.EnclaveIdentifier)
+		return nil, notFoundErr
```
will it return a 404 status code? shouldn't it be a "ResponseInfo" type?

leoporoli replied:
I see that the enclave handler is returning the "ResponseInfo" which specifies the status code but this handler isn't. shouldn't it return it?

lostbean replied:
Yes, I've also added them to engine api

leoporoli commented on engine/server/engine/server/engine_rest_api_handler.go:
```diff
@@ -0,0 +1,279 @@
+	case types.EnclaveContainersStatus_RUNNING:
+		return api_type.EnclaveContainersStatusRUNNING
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", status))
```
propagate error instead of panicking here

leoporoli commented on engine/server/engine/server/engine_rest_api_handler.go:
```diff
@@ -0,0 +1,279 @@
+	case types.ContainerStatus_RUNNING:
+		return api_type.ApiContainerStatusRUNNING
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", status))
```
same here

leoporoli commented on engine/server/engine/server/engine_rest_api_handler.go:
```diff
@@ -0,0 +1,279 @@
+	case types.EnclaveMode_TEST:
+		return api_type.TEST
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", mode))
```
same here

leoporoli commented on engine/server/engine/server/engine_rest_api_handler.go:
```diff
@@ -0,0 +1,279 @@
+	case types.EnclaveContainersStatus_EMPTY:
+		return api_type.EnclaveContainersStatusEMPTY
+	default:
+		panic(fmt.Sprintf("Undefined mapping of value: %s", status))
```
same here

leoporoli commented on engine/server/engine/server/websocket_api_handler.go:
```diff
@@ -0,0 +1,386 @@
+	MetricsClient metrics_client.MetricsClient
+}
+
+func sendErrorCode(ctx echo.Context, code int, message string) error {
```
we usually put private functions after public ones

leoporoli commented on engine/server/engine/server/websocket_api_handler.go:
```diff
@@ -0,0 +1,386 @@
+		params.ConjunctiveFilters,
+	)
+	if err != nil {
+		logrus.Error(err)
```
I think we could add more context in this error, "Failed to get log streamer with error:" for instance

leoporoli commented on engine/server/engine/server/websocket_api_handler.go:
```diff
@@ -0,0 +1,386 @@
+	enclaveUuid, err := engine.EnclaveManager.GetEnclaveUuidForEnclaveIdentifier(context.Background(), enclaveIdentifier)
+	if err != nil {
+		logrus.Errorf("An error occurred while fetching uuid for enclave '%v'. This could happen if the enclave has been deleted. Treating it as UUID", enclaveIdentifier)
+		return nil, err
```
missing propagate

leoporoli commented on engine/server/engine/server/websocket_api_handler.go:
```diff
@@ -0,0 +1,386 @@
+
+	notFoundServiceUuids, err := engine.reportAnyMissingUuidsAndGetNotFoundUuidsListHttp(context, enclaveUuid, requestedServiceUuids)
+	if err != nil {
+		return nil, err
```
missing propagate, I suggest to check them  throughout the file

leoporoli commented on engine/server/engine/server/websocket_api_handler.go:
```diff
@@ -0,0 +1,386 @@
+	return nil
+}
+
+func (streamer LogStreamer) streamHTTP(ctx echo.Context) error {
```
nice!!

leoporoli reviewed (commented):
approved with some comments

lostbean committed:
Add async starlark logs

lostbean committed:
Merge OpenAPI specs

lostbean committed:
Fix create enclave optional params and keep ctx in background

lostbean committed:
add async logs to websocket api

lostbean committed:
Use LRU with TTL to manage logs streamers

lostbean committed:
move code into better modules

lostbean committed:
add missing Error conversion

lostbean committed:
return after reply

lostbean committed:
fix return values

lostbean committed:
remove deprecated field from rest api

lostbean committed:
make improvements to the API names and descriptions

lostbean committed:
Improve main engine function

lostbean committed:
improvements to the enclave handler module

lostbean committed:
clean-up utils

lostbean committed:
report existing file artifact upload

lostbean committed:
do not panic

lostbean committed:
add exhaustive linter

lostbean committed:
improve responses on engine api

lostbean committed:
improve websocket logs

lostbean committed:
remaining golint errors

lostbean committed:
Merge remote-tracking branch 'origin/main' into edgar/unified_restful_engine

lostbean committed:
fix lint errors in container-engine-lib due new linter

lostbean committed:
fix lint errors in grpc-file-transfer-lib due new linter

lostbean committed:
bump go in CI lint

lostbean committed:
set go module off during linter CI

lostbean committed:
tidy it all

lostbean committed:
use quote to force version to be string

lostbean committed:
fix another missing enum case

lostbean committed:
fix linter errors in other modules

lostbean committed:
Merge branch 'main' into edgar/unified_restful_engine

lostbean committed:
update all references to go 1.20

lostbean committed:
move from math/rand to crypto/rand

rand.Seed has been deprecated since Go 1.20 and an alternative has been
available since Go 1.0: Programs that call Seed and then expect a
specific sequence of results from the global random source (using functions such as Int)
can be broken when a dependency changes how much it consumes from the
global random source.

To avoid such breakages, programs that need a specific result
sequence should use NewRand(NewSource(seed))
to obtain a random generator that other packages cannot access

lostbean committed:
add script to run go linter on all go modules

lostbean committed:
add script to tidy all go modules

lostbean committed:
port number typo

lostbean committed:
Merge branch 'main' into edgar/unified_restful_engine
